File: go-stream/offset_tracking_send.go
package main

import (
	"fmt"
	"github.com/rabbitmq/rabbitmq-stream-go-client/pkg/amqp"
	"github.com/rabbitmq/rabbitmq-stream-go-client/pkg/stream"
	"os"
)

func main() {
	env, err := stream.NewEnvironment(
		stream.NewEnvironmentOptions().
			SetHost("localhost").
			SetPort(5552).
			SetUser("guest").
			SetPassword("guest"))
	CheckErrSend(err)

	streamName := "stream-offset-tracking-go"
	err = env.DeclareStream(streamName,
		&stream.StreamOptions{
			MaxLengthBytes: stream.ByteCapacity{}.GB(2),
		},
	)
	CheckErrSend(err)

	producer, err := env.NewProducer(streamName, stream.NewProducerOptions())
	CheckErrSend(err)

	messageCount := 100
	chPublishConfirm := producer.NotifyPublishConfirmation()
	ch := make(chan bool)
	handlePublishConfirm(chPublishConfirm, messageCount, ch)

	fmt.Printf("Publishing %d messages...\n", messageCount)
	for i := 0; i < messageCount; i++ {
		var body string
		if i == messageCount-1 {
			body = "marker"
		} else {
			body = "hello"
		}
		err = producer.Send(amqp.NewMessage([]byte(body)))
		CheckErrSend(err)
	}
	_ = <-ch
	fmt.Println("Messages confirmed.")

	err = producer.Close()
	CheckErrSend(err)
}

func handlePublishConfirm(confirms stream.ChannelPublishConfirm, messageCount int, ch chan bool) {
	go func() {
		confirmedCount := 0
		for confirmed := range confirms {
			for _, msg := range confirmed {
				if msg.IsConfirmed() {
					confirmedCount++
					if confirmedCount == messageCount {
						ch <- true
					}
				}
			}
		}
	}()
}

func CheckErrSend(err error) {
	if err != nil {
		fmt.Printf("%s ", err)
		os.Exit(1)
	}
}

File: javascript-nodejs-stream/offset_tracking_send.js
const rabbit = require("rabbitmq-stream-js-client");

async function main() {
  console.log("Connecting...");
  const client = await rabbit.connect({
    vhost: "/",
    port: 5552,
    hostname: "localhost",
    username: "guest",
    password: "guest",
  });

  console.log("Making sure the stream exists...");
  const streamName = "stream-offset-tracking-javascript";
  await client.createStream({ stream: streamName, arguments: {} });

  console.log("Creating the publisher...");
  const publisher = await client.declarePublisher({ stream: streamName });

  const messageCount = 100;
  console.log(`Publishing ${messageCount} messages`);
  for (let i = 0; i < messageCount; i++) {
    const body = i === messageCount - 1 ? "marker" : `hello ${i}`;
    await publisher.send(Buffer.from(body));
  }

  console.log("Closing the connection...");
  await client.close();
}

main()
  .then(() => console.log("done!"))
  .catch((res) => {
    console.log("Error in publishing message!", res);
    process.exit(-1);
  });
